library/core/src/iter/adapters/array_chunks.rs

```diff
@@ -0,0 +1,146 @@
+use core::iter::FusedIterator;
+use core::mem::{self, MaybeUninit};
+use core::ptr;
+
+/// An iterator that yields the elements of another iterator in
+/// chunks of size `N`.
+///
+/// This `struct` is created by the [`array_chunks`] method on [`Iterator`]. See
+/// its documentation for more.
+///
+/// [`array_chunks`]: Iterator::array_chunks
+#[unstable(feature = "iter_array_chunks", issue = "none")]
+#[derive(Debug)]
+pub struct ArrayChunks<I: Iterator, const N: usize> {
+    iter: I,
+    buffer: [MaybeUninit<I::Item>; N],
+    init: usize,
+}
+
+#[unstable(feature = "iter_array_chunks", issue = "none")]
+impl<I: Iterator, const N: usize> Drop for ArrayChunks<I, N> {
+    fn drop(&mut self) {
+        // SAFETY: This is safe: `remainder_mut` returns exactly the sub-slice
+        // of elements that were initialized but not yielded and so have yet
+        // to be dropped.
+        unsafe {
+            ptr::drop_in_place(self.remainder_mut());
+        }
+    }
+}
+
+impl<I: Iterator, const N: usize> ArrayChunks<I, N> {
+    pub(in crate::iter) fn new(iter: I) -> Self {
+        Self { iter, init: 0, buffer: MaybeUninit::uninit_array() }
+    }
+
+    /// Returns the remainder of the elements yielded by the original
+    /// iterator that were insufficient to fill another chunk. The
+    /// returned slice has at most `N-1` elements.
+    #[unstable(feature = "iter_array_chunks", issue = "none")]
+    pub fn remainder(&self) -> &[I::Item] {
+        // SAFETY: We know that all elements before `init` are properly initialized.
+        unsafe { MaybeUninit::slice_assume_init_ref(&self.buffer[..self.init]) }
+    }
+
+    /// Returns the remainder of the elements yielded by the original
+    /// iterator that were insufficient to fill another chunk. The
+    /// returned slice has at most `N-1` elements.
+    #[unstable(feature = "iter_array_chunks", issue = "none")]
+    pub fn remainder_mut(&mut self) -> &mut [I::Item] {
+        // SAFETY: We know that all elements before `init` are properly initialized.
+        unsafe { MaybeUninit::slice_assume_init_mut(&mut self.buffer[..self.init]) }
+    }
+}
+
+#[unstable(feature = "iter_array_chunks", issue = "none")]
+impl<I: Iterator, const N: usize> Iterator for ArrayChunks<I, N> {
```

Comment:
The implementation should at least get `size_hint` and `advance_by`. `fold` and `try_fold` could also make a big difference for efficiency.

Once `size_hint` is correctly implemented a conditional `TrsutedLen` impl could be added.

You can look at other Iterator adapters as a reference.

Reply:
I don't know if this iterator can actually uphold the contract of `TrustedLen`. If the underlying iterator returns `(usize::MAX, None)`, in most cases it would not be possible to tell whether this iterator would return less than or equal to `usize::MAX` elements (arrays). For example, with `N=2` and actual length of the underlying iterator being `usize::MAX+1`, it would be less than `usize::MAX`, but with `N=2` and actual length of the underlying iterator being `10*usize::MAX`, it would be more than `usize::MAX`. I think the only case where it would be possible is when `N=1`.

Reply:
Ok, I have added those methods now. The `advance_by` method is potentially lossy though if the underlying iterator advances by less than the requested amount.

Reply:
> I think the only case where it would be possible is when N=1.

Good point. No need to implement it then, arrays of len 1 don't make much sense.

>  The advance_by method is potentially lossy though

Ah that would break the contract about the remainder. Then can only be implemented when the size is known, i.e. only via specialization. Or maybe by querying the lower size bound. Since it's only a performance optimization we can leave it out for now.

Reply:
But I think it can implement `ExactSizeIterator` when the inner is since that means the inner fits in usize::MAX and thus we can calculate the hint properly.